src/backend/distributed/executor/multi_utility.c

```diff
@@ -3787,3 +3804,78 @@ ProcessDropTableStmt(DropStmt *dropTableStatement)
 		}
 	}
 }
+
+
+/*
+ * ShouldExecuteAlterTableSequentially checks if the given ALTER TABLE
```

Comment:
I think we should put all the commands that need sequential execution into here. So, we should move the logic below to here, otherwise, it'd be very hard to follow when we switch to seq, mode

```
		/*
		 * We need to execute the ddls working with reference tables on the
		 * right side sequentially, because parallel ddl operations
		 * relating to one and only shard of a reference table on a worker
		 * may cause self-deadlocks.
		 */
		if (PartitionMethod(rightRelationId) == DISTRIBUTE_BY_NONE)
		{
			ddlJob->executeSequentially = true;
		}
```

I haven't looked into detail on how to move that part here, but, feel free to discuss it is hard to implement.

(Wild guess: Is it simply checking for `AT_AddConstraint` ?)